### PR TITLE
8327478: Add System test to verify TextSelection issue for webkit-617.1

### PR DESCRIPTION
**Pull request by Gopalora**
There was no test included with the fix for [JDK-8326989](https://bugs.openjdk.org/browse/JDK-8326989),

Hence we are adding a system test now.

Test is written as 
1. Load html content in web view.
2. pick the color of mouse pointer.
3. Perform double click.
4. pick the color again.
> expected bahaviour: colour picked in step 2 and 4 should not match.

Verification:
The test passes with latest webkit source
Also verified that test fails when the fix for [JDK-8326989](https://bugs.openjdk.org/browse/JDK-8326989) is reverted.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327478](https://bugs.openjdk.org/browse/JDK-8327478): Add System test to verify TextSelection issue for webkit-617.1 (**Bug** - P3)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1719/head:pull/1719` \
`$ git checkout pull/1719`

Update a local copy of the PR: \
`$ git checkout pull/1719` \
`$ git pull https://git.openjdk.org/jfx.git pull/1719/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1719`

View PR using the GUI difftool: \
`$ git pr show -t 1719`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1719.diff">https://git.openjdk.org/jfx/pull/1719.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1719#issuecomment-2691547461)
</details>

### CODE REVIEW

**Comment by kevinrushforth** [comment]
@Gopalora Please enable GHA tests on your repo.

I'll provide a few comments now, and formally review it once your OCA status is verified.

**Comment by Gopalora** [comment]
Thanks for the review.
I have done the code change as per the suggestions in the code file.
Please have a relook of the changes.

File: tests/system/src/test/java/test/util/Util.java [hunk 1]
```diff
@@ -333,6 +334,20 @@ public static void waitForLatch(CountDownLatch latch, int seconds, String msg) {
         }
     }
 
+    /**
+     * Makes double click of the mouse left button.
+     */
+    public static void doubleClick(Robot robot, int x, int y) {
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/util/Util.java:340]: Since this is now a general-purpose utility, I'd prefer to separate out the mouse move from the double-click (with the mouse move being done in the test itself) and just have this be `void doubleClick(Robot robot)`.

Comment (Gopalora) [reply tests/system/src/test/java/test/util/Util.java]: Done.

**Comment by kevinrushforth** [comment]
The latest changes look good. I'll formally review it once your OCA status is verified.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+import test.util.Util;
+
+public class TextSelectionTest {
+
+    private static final String html = "<html>" +
```
Comment (andy-goryachev-oracle) [inline tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:47]: we could use new java text block here

```
private static final String html =
    """
    <html>
    <head></head>
    <body>&nbsp&nbsp&nbsp&nbsp some text</body>
    </html>
    """;
```

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import javafx.stage.Stage;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+import test.util.Util;
+
+public class TextSelectionTest {
```
Comment (andy-goryachev-oracle) [inline tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:45]: I wonder if we could extend RobotTestBase base class to simplify the code?

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        Util.shutdown();
+    }
+
+    // ========================== TEST CASE ==========================
+    @Test
+    @Timeout(value=20)
+    public void testTextSelection() throws Exception {
+
+        int x = (int)(scene.getWindow().getX() + scene.getX() + 22);
+        int y = (int)(scene.getWindow().getY() + scene.getY() + 15);
```
Comment (andy-goryachev-oracle) [inline tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:97]: can these offsets (22, 15) be measured instead of hard-coded?

Comment (Gopalora) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:77]: We are not getting the text area from web view. This point is used to simulate the mouse click.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java (hunk 2 above)
Comment (Gopalora) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java]: Done.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java (hunk 3 above)
Comment (Gopalora) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java]: Done.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+
+import test.robot.testharness.RobotTestBase;
+import test.util.Util;
+
+public class TextSelectionTest extends RobotTestBase{
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:42]: Minor: please add a space before the `{`

**Review by kevinrushforth — COMMENTED** [review]
Looks good with one minor code style issue.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    private static final String html =
+    """
+    <html>
+    <head></head>
+    <body>&nbsp&nbsp&nbsp&nbsp some text</body>
+    </html>
+    """;
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:50]: One more minor code style issue: lines 45-50 should be intended

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:77]: so they are arbitrary, and will work for every font size?

(The default font size varies depending on the platform and other circumstances, see `PrismFontFactory::getSystemFontSize()` - though I am not sure if the default font size is picked up by the `WebView`.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+    private static CountDownLatch webviewLoadLatch = new CountDownLatch(1);
+    private Color colorBefore;
```
Comment (andy-goryachev-oracle) [inline tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:53]: suggestion: make `colorBefore/After` _volatile_ since they are being accessed from different threads without any synchronization.
I think this should be sufficient (no need for `AtomicReference`).

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        Util.runAndWait(() -> colorBefore = robot.getPixelColor(x, y));
+
+        Util.runAndWait(() -> robot.mouseMove(x, y));
+        Util.doubleClick(robot);
+        Util.sleep(500); // Wait for the selection highlight to be drawn
+
+        Util.parkCursor(robot);
+        Util.runAndWait(() -> colorAfter = robot.getPixelColor(x, y));
+
+        Assertions.assertNotEquals(colorBefore, colorAfter,
```
Comment (andy-goryachev-oracle) [inline tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:89]: should we also test for non-null `colorBefore` ?

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java (hunk 7 above)
Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:53]: https://en.wikipedia.org/wiki/Cache_coherence

Comment (kevinrushforth) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:53]: > https://en.wikipedia.org/wiki/Cache_coherence

That's interesting, but not particularly relevant. What is relevant is the JVM spec, specifically the Java Memory Model, which guarantees a happens-before ordering of actions in two threads that synchronize-with each other.

> edit: you might be right about runAndWait, we know it does separate the access from multiple threads in time, and it might be even smart enough to figure out that the field is being accessed from two different threads and emit the appropriate bytecode (similarly how we don't need the `final` keyword for effectively final variables nowadays).

It isn't about the byte code that is emitted as much as it is about how the code is executed by the two different threads. And it isn't because it need to be "smart enough", it's because that's how the memory model works.

A write by the thread that executes the runnable passed into runAndWait (the FX app thread in this case), happens-before the read done by the test thread that is done after the runAndWait returns because the runAndWait call causes the actions done in the runAndWait's runnable to synchronize-with the wait done by the calling thread.

> But I think that adding `volatile` **_guarantees_** that all the necessary steps are done by the JVM.

Yes, adding `volatile` guarantees the appropriate happens-before relationship. But so does synchronizing.

I don't really care one way or the other in this case.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java (hunk 8 above)
Comment (kevinrushforth) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:89]: `Robot::getPixelColor` can't return null (although I checked the docs and we don't specify one way or the other). It seems fine either way.

Comment (andy-goryachev-oracle) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:89]: what if the lambdas never run?

Comment (kevinrushforth) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:89]: Then we have far bigger problems than this one test. The reason we use `runAndWait` instead of `runLater` is so we guarantee that the lambda has run before that method returns. It's guaranteed to do so.

Comment (Gopalora) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java:87]: Added volatile to the variables colorBefore and colorAfter.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java (hunk 5 above)
Comment (Gopalora) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java]: Done.

File: tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java (hunk 6 above)
Comment (Gopalora) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java]: Done as per one suggestion on "https://docs.oracle.com/en/java/javase/18/text-blocks/index.html"

Comment (Gopalora) [reply tests/system/src/test/java/test/robot/javafx/web/TextSelectionTest.java]: Removed the line "<head></head>" as this is empty.

**Review by andy-goryachev-oracle — APPROVED** [review]
lgtm
